SEED-platform/seed
# Quantity (aka. Pint) fields to accommodate metric and US Customary energy/area data

#### Any background context you want to provide?

Use of SEED by Canadian organizations requires importing energy/area data as either imperial or metric (or, in the case of the City of Toronto, in imperial energy units but metric area units (!)). Different organizations tend to culturally work with different units as well: eg Passive House-influenced organizations will tend to use kWh/m²/year, where federal agencies favour GJ/m²/year.

The database groundwork for accommodating this was laid in #1457; this completes that work with minor database work and more user-interface work. Recognizing that there is quite a lot of existing production data this touches, we are favouring a conservative approach has of adding new EUI and area database columns to accommodate this, rather than changing existing ones.

#### What's this PR do?

1. Renames the original float-type fields for EUIs and areas to add the suffix `_orig`.
1. Renames the Quantity fields with suffix `_pint` that were introduced in #1457. It remove the suffix and these become the canonical columns for this data.
1. Migrates existing PropertyStates to bring forward old float data as new Quantity data, preserving old float data in case of catastrophe.
1. Migrates any existing data quality rules to have Pint-parseable units (eg. 'ft**2' instead of 'sq. ft'). 
1. Adds an organization-level preference for the units used in display of EUI and area data, including persistence and user interface.
1. Adds server-side code needed to "collapse" Quantity data so that as far as the front-end is concerned, it is still getting floats and these are converted to reflect the organization's preference for units.
1. Adds server-side code to handle and report unit-aware data quality checks.
1. Removes release flipper `release:use_pint` (from #1457). 
1. Adds new release flipper `release:orig_columns` to hide the display of `_orig` float data in the user interface while letting it remain in the database. (This should go away in the 2.4 cycle, with the scheduled deletion of the `_orig` columns).

(cosmetic changes):
1. Alphabetically sorts existing English, French translation file to make future changes more readily apparent.

#### How should this be manually tested?

Needs QA against a fresh database, and an existing production database clone. Procedure is similar for both:

1. Create a new data set
1. Add data quality rule to label anything over 3 GJ/m2/year as "High EUI" (this is equivalent to 264 kBtu/ft2/year and should catch some reports in the sample data file).
1. Import sample Portfolio Manager file, proceed to mapping
1. Confirm "Site Eui" and "Source Eui" are showing `kBtu/sq ft/year` in unit selection
1. Map "Weather Normalized Site EUI (kBtu/ft2)" as "Site Eui Weather Normalized" and confirm that units dropdown has appeared, is showing "kBtu/ft2/year". (Repeat with the weather normalized source EUI).
1. Map file. Confirm that data quality results contain entries for "High EUI" (ie. that units have converted correctly between the metric rule and the US Customary data)
1. View inventory. EUIs should be showing in kBtu/m2/year.
1. Go to Organizations > (Name of Logged in Org) > Settings. Change EUI and Area units
1. Go back to inventory, confirm that inventory is now showing and exporting in new units, and that column headers are displaying correct units.
1. Go to the detail of a property, confirm that property states are showing correct unit data and that header is showing correct units.
1. Go to inventory report, confirm that EUI graphs are showing correct units in axis labels and that data is converted to correct units.

#### What are the relevant tickets?

#1457, (not sure otherwisen; don't have access to Pivotal tickets)

#### Screenshots (if appropriate)
<img width="1478" alt="org-prefs" src="https://user-images.githubusercontent.com/6805/34958327-370200c0-f9e6-11e7-8bdf-4279c2e8e361.png">
<img width="1396" alt="data-quality" src="https://user-images.githubusercontent.com/6805/34958326-36eacbbc-f9e6-11e7-8a64-45ee3ec7ed53.png">
<img width="1383" alt="units-mapping" src="https://user-images.githubusercontent.com/6805/34958330-3737700c-f9e6-11e7-9838-0f43ce154d63.png">
<img width="1478" alt="data-quality-result" src="https://user-images.githubusercontent.com/6805/34958325-36d3c444-f9e6-11e7-8ce4-30ee39aab951.png">
<img width="1478" alt="units-inventory" src="https://user-images.githubusercontent.com/6805/34958329-3721d79c-f9e6-11e7-84ba-88ce1bf15120.png">
<img width="1478" alt="units-reports" src="https://user-images.githubusercontent.com/6805/34958331-3753b334-f9e6-11e7-9c13-cc35c299e980.png">

#### Definition of Done:


No, something is definitely breaking within the test suite while manual testing in the browser seems OK. I'm having trouble telling where the problem is (maybe fixture data, assumptions that have become embedded within the test code, or genuine application-code bugs). I'd appreciate some assistance in diagnosing this.


May require a minor add to the API test to ensure that unit selection is correctly added to the mapping step


Don't believe so.


No.
